### PR TITLE
ubertest: non-FI_RX_CQ_DATA support, proper syncing

### PR DESCRIPTION
**Pull request by aingerson**
- add no FI_RX_CQ_DATA support to ubertest (to test shm provider properly)
- fix ubertest syncing method
- set default mode to FT_MODE_ALL
- update sockets config to include extra selective completion test cases

### CODE REVIEW

**Comment by aingerson** [comment]
@a-ilango see util_ep changes included here

**Comment by shefty** [comment]
From one of the commit messages:
```
utility
- add util_ep->flags to store whether the rx and tx queues are bound with selective flags for msg functions with or without the default op flags
 This is specifically to handle the case where:
 - ep_bind with SELECTIVE_COMPLETION
 - op_flags = FI_COMPLETION
 - msg_flags = 0
The endpoint needs to know whether to use the default op_flags for sendmsg/recvmsg if the expected behavior is suppression of the completion event.
(rxm needs to be upated to handle this)
```
The flags passed into sendmsg/recvmsg are always used in place of the default flags.  The default flags are used when flags are not passed as part of an API call.

**Comment by a-ilango** [comment]
> The flags passed into sendmsg/recvmsg are always used in place of the default flags. The default flags are used when flags are not passed as part of an API call.

When you say "flags are not passed" does it mean no flags are passed (flags=0)? If the user passes some flags and ~FI_COMPLETION, should the call not generate a completion? This pertains to the case where user doesn't pass FI_SELECTIVE_COMPLETION at bind time and tx op flags are unset (example 3 in the fi_ep_bind->FI_SELECTIVE_COMPLETION in fi_endpoint manpage).

**Comment by shefty** [comment]
"Flags are not passed" means that there is no flags argument at all.  Flags = 0 means that flags are being passed and that none are set.

**Comment by shefty** [comment]
Example 3 indicates that the provider will always generate a completion.  I.e. selective completions are disabled.  The FI_COMPLETION flag is ignored.

**Comment by aingerson** [comment]
Updated to include a man page update and also the fix you pointed out @shefty I'm looking into the ubertest issue with verbs
@a-ilango your rxm code is fine. APPARENTLY it's just my code that was wrong #thanksMark

**Comment by aingerson** [comment]
Re-added the util_ep fix. The other option is to add 2 fields to the util_ep to store the default completion flag and OR that in regardless of the binding flag. Thoughts @shefty?
Also jk @a-ilango I think rxm needs to handle this case after all. Also maybe tcp @vkrishna

**Comment by aingerson** [comment]
FYI I have a couple more patches that add the util_ep->cq_bind_flags to rxm, fix the verbs mode bit in ubertest, and a shm attr fix but I'm waiting until #4772 is merged to add those patches in since nothing is going to pass until it is.

**Comment by aingerson** [comment]
I've taken out all the shm changes and opened a separate pull request for those. What started out as a simple shm addition turned into a big ubertest project....
See #4781 for those changes
@a-ilango see the new endpoint sync in ubertest

**Comment by aingerson** [comment]
Investigating mlx hang

**Comment by aingerson** [comment]
I can't reproduce the hang at all and retested it 5 times on the CI but it's not happening again so I'm not sure what that's all about...

File: fabtests/ubertest/cq.c [hunk 1]
```diff
@@ -253,10 +251,30 @@ int ft_bind_comp(struct fid_ep *ep)
...
+	size_t credits = 0;
+	uint64_t flags;
+
+	if (format == FI_CQ_FORMAT_CONTEXT || test_info.mode & FI_RX_CQ_DATA)
+		return count;
+
+	while (count--) {
+		flags = ((struct fi_cq_msg_entry *) ptr)->flags;
+		if (flags & FI_TRANSMIT || !(flags & FI_REMOTE_CQ_DATA && flags & FI_RMA))
+			credits++;
```
[the remark this answers is not in the document]
Comment (aingerson) [reply fabtests/ubertest/cq.c]: Good call. I'll update it. That gives me another chance to maybe recreate that mlx hang. Any idea what could have caused that?

[the remark this answers is not in the document]
Comment (a-ilango) [reply fabtests/ubertest/cq.c]: Which test was hanging?

[the remark this answers is not in the document]
Comment (aingerson) [reply fabtests/ubertest/cq.c]: bw send()

**Comment by shefty** [comment]
Is this PR okay to merge separately, or is it still undergoing development?

**Comment by aingerson** [comment]
I think it should be fine to merge. There was one hang with verbs (mlx) that happened once but I can't reproduce it on my own and no CI rebuilds seem to reproduce it either so I'm not sure what that is all about. Your call

**Pull request merged**